### PR TITLE
bugfix: LB_ORIGINAL_DST retryTime is 0

### PR DESCRIPTION
**Pull request by 3062**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 1]
```diff
@@ -147,7 +147,7 @@ func (p *proxy) initializeUpstreamConnection() api.FilterStatus {
 	}
 
 	retryTime := clusterSnapshot.HostSet().Size()
-	if retryTime > defaultConnectRetryTimes {
+	if retryTime > defaultConnectRetryTimes || (clusterSnapshot.ClusterInfo().LbType() == types.ORIGINAL_DST && retryTime == 0) {
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:150]: 分成两个if 吧，清晰一些
并且retrytime为0，且不是origin也需要打印error日志

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 2]
```diff
@@ -150,6 +150,13 @@ func (p *proxy) initializeUpstreamConnection() api.FilterStatus {
 	if retryTime > defaultConnectRetryTimes {
 		retryTime = defaultConnectRetryTimes
 	}
+	if retryTime == 0 {
+		if clusterSnapshot.ClusterInfo().LbType() == types.ORIGINAL_DST {
+			retryTime = defaultConnectRetryTimes
+		} else {
+			log.DefaultLogger.Errorf("%s proxy connect retryTime is 0", p.network)
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:157]: We should log more cluster info.

**Review by taoyuanyuan — DISMISSED** [review]
LGTM

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 3]
```diff
@@ -150,6 +150,13 @@ func (p *proxy) initializeUpstreamConnection() api.FilterStatus {
 	if retryTime > defaultConnectRetryTimes {
 		retryTime = defaultConnectRetryTimes
 	}
+	if retryTime == 0 {
+		if clusterSnapshot.ClusterInfo().LbType() == types.ORIGINAL_DST {
```
Comment (nejisama) [inline pkg/filter/network/streamproxy/streamproxy.go:154]: 这里为什么一定要限制 ORIGINAL_DST  是不是任意都可以？

Comment (3062) [reply pkg/filter/network/streamproxy/streamproxy.go:154]: ORIGINAL_DST 一般不配置 hosts 所以这个值为 0 是正常的，其余情况为 0 需要报 error

Comment (nejisama) [reply pkg/filter/network/streamproxy/streamproxy.go:154]: 1. cluster type 也有一个ORIGINAL_DST 不能用这个吗。
2. 另外我看https://github.com/mosn/mosn/blob/master/pkg/upstream/cluster/original_dst.go 的Number 总是返回1的，这个0 是怎么来的，是不是不太对

Comment (nejisama) [reply pkg/filter/network/streamproxy/streamproxy.go:154]: 看了下 应该是clusterSnapshot.HostSet().Size() 这个接口调用不太对
如果改成 clusterSnapshot.LoadBalancer(). HostNum 好像就可以了，似乎用这个更合理一点

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 4]
```diff
@@ -146,17 +146,10 @@ func (p *proxy) initializeUpstreamConnection() api.FilterStatus {
 		cluster: clusterInfo,
 	}
 
-	retryTime := clusterSnapshot.HostSet().Size()
+	retryTime := clusterSnapshot.LoadBalancer().HostNum(nil)
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:149]:
```suggestion
	retryTime := clusterSnapshot.HostNum(nil)
```

File: pkg/filter/network/streamproxy/streamproxy.go [hunk 5]
```diff
@@ -146,7 +146,7 @@ func (p *proxy) initializeUpstreamConnection() api.FilterStatus {
...
```
Comment (taoyuanyuan) [inline pkg/filter/network/streamproxy/streamproxy.go:149]: you should log error when retryTime equal 0.

**Review by taoyuanyuan — APPROVED** [review]

**Review by nejisama — APPROVED** [review]

**Pull request merged**